internal/plugin/metrics_plugin.go

```diff
@@ -0,0 +1,257 @@
+	// Initialize the first time only if the Prometheus config section is configured.
+	if m.conf.Metrics.PrometheusSource == nil {
+		slog.Debug("Prometheus metrics source not configured: skipping initialization")
+		return
+	}
+
+	if _, ok := m.producers[model.Prometheus]; !ok {
+		m.producers[model.Prometheus] = prometheus.NewScraper(m.conf.Metrics.PrometheusSource.Endpoints)
+	} else {
+		slog.Debug("Prometheus metrics source already initialized")
+	}
+}
+
+func (m *Metrics) createExporters(ctx context.Context) error {
+	if m.conf.Metrics.OTelExporter == nil {
+		slog.Debug("OTel exporter not configured: skipping exporter instantiation")
+		return nil
+	}
+
+	if _, ok := m.exporters[model.OTel]; !ok {
+		exporter, err := export.NewOTelExporter(
+			ctx, m.conf, model.Prometheus.String(), prometheus.ConvertPrometheus,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to create OTel exporter: %w", err)
+		}
+
+		m.exporters[exporter.Type()] = exporter
+	}
+
+	return nil
+}
+
+func (m *Metrics) startExporters(ctx context.Context) {
+	for expType, exp := range m.exporters {
+		slog.Info("Starting export", "producer_type", expType.String())
+		go exp.StartSink(ctx)
+	}
+}
+
+// Produces metrics from the MetricsProducer at the configured interval and pushes to the given channel.
+func (m *Metrics) runProducer(ctx context.Context, producer model.MetricsProducer) {
+	t := time.NewTicker(m.collectInterval)
+
+	failedAttempts := 0
+	for failedAttempts != maxProducerRetries {
+		select {
+		case <-t.C:
+			failedAttempts = m.callProduce(ctx, producer, failedAttempts)
+		case <-ctx.Done():
+			slog.Info("MetricsProducer stopped", "producer_type", producer.Type().String())
+
+			return
+		}
+	}
+
+	if failedAttempts == maxProducerRetries {
+		slog.Error("MetricsProducer stopped after max number of retries reached",
+			"producer_type", producer.Type().String(), "max_retries", maxProducerRetries)
+	}
+}
+
+func (m *Metrics) processMessage(msg *bus.Message) {
+	de, ok := msg.Data.(model.DataEntry)
+	if !ok {
+		slog.Error("Metrics plugin received metrics event but could not cast it to correct type",
+			"payload", msg.Data)
+
+		return
+	}
+
+	exporter, ok := m.exporters[model.OTel]
+	if !ok {
+		slog.Error("Metrics plugin received metrics event but source type had no exporter",
+			"source_type", de.SourceType)
+	} else {
+		err := exporter.Export(de)
+		if err != nil {
+			slog.Error("failed to export metrics to data sink", "error", err)
+		}
+	}
+}
+
+func (m *Metrics) callProduce(ctx context.Context, producer model.MetricsProducer, failedAttempts int) int {
+	entries, err := producer.Produce(ctx)
+	if err != nil {
+		slog.Debug("MetricsProducer produce call failed", "error", err)
+
+		return failedAttempts + 1
+	}
+	failedAttempts = 0
+
+	busMsgs := make([]*bus.Message, len(entries))
+	for i, e := range entries {
+		busMsgs[i] = e.ToBusMessage()
+	}
+
+	m.pipe.Process(busMsgs...)
+
+	return failedAttempts
+}
+
+//##########################
```

Comment:
Not sure about having these MetricsOptions comments here

Reply:
True, removing